95.md

```diff
@@ -0,0 +1,91 @@
+NIP-95
+======
+
+File Storage
+------------
+
+`draft` `optional` `author:arthurfranca` `author:Semisol` `author:frbitten` `author:staab`
+
+## Motivation
+
+This NIP defines a way to turn a relay into a storage service.
+The spec doesn't use regular nostr events through websockets for
+storing, requesting nor retrieving data because it would be inneficient.
+
+## Relay Adaptation
+
+Relays wishing to be used as file storage should opt-in by making available
+an https route at `/nip95(/<sha256-file-hash>)` for handling files as described below.
+They may announce support with [NIP-11](11.md) `supported_nips` field.
+Other ways of advertising NIP-95 support may be described later.
+
+## Upload
+
+A file can be uploaded one at a time to `https://relay.domain/nip95` as `multipart/form-data` content type using `POST` method with the file object set to the `file` form data field.
```

Comment:
If you take a look at dufflepud (coracle's helper server), you'll see a [quote endpoint](https://github.com/coracle-social/dufflepud/blob/44660fb2398cc56e67f3d795cb141e7f817c52fa/dufflepud/app.py#L71) that allows the client to negotiate the file upload, for example generating an invoice for the user to pay prior to upload based on file expiration, size, or other information. I'd like to see relays compensated for this kind of heavy data storage. This also provides a lightweight extension mechanism for future use cases that would be hard to implement without a negotiation step.

Reply:
I agree. We should add expiration (empty if no expiration) and size fields and the error message should tell what should be paid or went wrong. Or do you envision a more involved negotiation?

Reply:
I was thinking of using a separate endpoint, but I think error messages would also work fine as long as servers don't read the full upload before returning errors. With multipart/form-content this might be hard to avoid though if form fields come after the file body. I'm not sure about that though.

We could also do something like allow the user to publish the header object and return OK false if rejected for whatever reason. The relay would then hold on to that event until the file is actually posted. More complicated, but fewer APIs to specify.

Reply:
> hard to avoid though if form fields come after the file body

Now I'm not sure I'll look into it

Reply:
The order is important (it is also true when uploading directly to AWS S3). So the `file` field must be the last one, or be reordered to the last position before submission or use XHR2's FormData to append the file field last. I will add this notice to the PR. Optionally the server may save the file to temp folder while the other fields weren't received but I wouldn't go that route.

So as AWS S3 is the king and this is a limitation of multipart/form-data, no problem. Agree?